Make random_sample() reproducible

wingkitlee0 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

wingkitlee0 commented:
@alexeykudinkin can you help review? thanks

alexeykudinkin commented:
Will be addressed in https://github.com/ray-project/ray/pull/46088

wingkitlee0 commented:
> Will be addressed in #46088

@alexeykudinkin I am aware of that PR and do not think the solution is correct. See my example here: https://github.com/ray-project/ray/pull/49443#issuecomment-2629119794

raulchen commented:
hey @wingkitlee0 , @alexeykudinkin was referring to a different PR https://github.com/ray-project/ray/pull/46088/files
That PR looks good to me as well. And it's simpler. 
If you still have concerns, please comment on that PR.

raulchen commented:
Actually, after reading more on both PRs, I think this PR is better, because it can avoid the small-batch issue mentioned in https://github.com/ray-project/ray/pull/46088 cc @alexeykudinkin

raulchen commented on python/ray/data/tests/test_map.py:
```diff
@@ -1551,6 +1551,100 @@ def test_random_sample_checks(ray_start_regular_shared):
+    # Expected:
+    expected = get_expected_mask_indices(
+        num_blocks=10, num_rows_per_block=10, fraction=0.1, seed=1234
+    )
```
we can probably just check against hard-coded expected results. 
otherwise, we'll need to also test get_expected_mask_indices is deterministic.

raulchen commented on python/ray/data/tests/test_map.py:
```diff
@@ -1551,6 +1551,100 @@ def test_random_sample_checks(ray_start_regular_shared):
+@pytest.mark.parametrize("seed", [1234, 4321, 0])
+def test_random_sample_fixed_seed_0002(
+    ray_start_regular_shared, dtype, num_blocks, num_rows_per_block, fraction, seed
+):
```
nice test

alexeykudinkin commented on python/ray/data/_internal/logical/operators/map_operator.py:
```diff
@@ -69,6 +69,7 @@ def __init__(
         fn_constructor_args: Optional[Iterable[Any]] = None,
         fn_constructor_kwargs: Optional[Dict[str, Any]] = None,
         min_rows_per_bundled_input: Optional[int] = None,
+        include_task_idx: bool = False,
```
Let's avoid this param

wingkitlee0 replied:
Sure. 
However, it will now need to call `_map_batches_without_batch_size_validation` which does not have default args. So there will be a bunch of hardcoded default values in `random_sample`.. Any advice?

alexeykudinkin commented on python/ray/data/dataset.py:
```diff
@@ -1580,26 +1593,29 @@ def random_sample(
+            if seed is None:
+                rng = np.random.default_rng()
+            else:
+                rng = np.random.default_rng([batch_idx, task_idx, seed])
```
Why do we need batch_idx?

wingkitlee0 replied:
My original thought was to keep this a pure and stateless function.

alexeykudinkin commented:
@wingkitlee0 yes, you're right. I overlooked the fact that we'll be generating identical sequences per block which isn't ideal.

wingkitlee0 commented on python/ray/data/dataset.py:
```diff
@@ -555,6 +555,8 @@ def __call__(self, batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
                 avoid "assignment destination is read-only" or "buffer source array is
                 read-only" errors. Default is ``False``.
+            include_task_idx: Whether to pass a task index to `fn`,
+                i.e., fn(task_idx, batch_idx, batch)
```
> 1) use a class-based UDF

class-based UDF requires `concurrency` to be set. Any way to get around that?

raulchen replied:
You can use a class-based UDF with `compute=TaskPoolComputeStrategy`

wingkitlee0 replied:
Interesting, `ray/data/_internal/util.py`'s `get_compute_strategy` seems to discourage the use of `CallableClass`  + `TaskPoolStrategy` (`get_compute_strategy` is called by `_map_batches_without_batch_size_validation`). I can find ways to bypass that, but I am curious if that would work...

raulchen replied:
By default, we choose ActorPoolStrategy if the UDF is a class. 
The motivation is to simplify the usage and avoid users having to specify the strategy. 
But in theory, CallableClass + TaskPoolStrategy should also be feasible.
You can probably move get_compute_strategy out of  _map_batches_without_batch_size_validation. 
Not sure if there are other issues. if you find you have to tweak too many things, TaskContext.kwargs is also fine.

wingkitlee0 commented:
It's ready for re-review!

* No change to the public APIs.
* Used `TaskContext.kwargs` at the end. I checked `TaskPoolStrategy` etc: It can't use class-based UDFs now because it misses the step to call the constructor.
* Updated AbstractUDFMap to use kwargs (the use of positional args + defaults makes it hard to track down the previous pipeline failures: the object was instantiated successfully with mismatched args but the error wasn't raised many steps later)
* Simplified the unit tests

raulchen commented on python/ray/data/dataset.py:
```diff
@@ -1582,26 +1587,39 @@ def random_sample(
+            else:
+                rng = np.random.default_rng(
+                    [ctx.kwargs.get("batch_idx", 0), ctx.task_idx, seed]
+                )
```
I think we can get rid of the `include_task_ctx` flag as well. 
1. we can add a thread-local variable in `TaskContext` to allow accessing the current TaskContext. Basically call `TaskContext.set_current/reset_currernt`  in `_map_task`. 
2. I don't think we need batch_idx here. We just need to create the rng object once per task and store it in `TaskContext.kwargs`.

wingkitlee0 replied:
Great suggestion. It will reduce code changes. I will probably need to update the tests a little bit.

raulchen commented on python/ray/data/_internal/execution/interfaces/task_context.py:
```diff
@@ -42,3 +46,36 @@ class TaskContext:
     kwargs: Dict[str, Any] = field(default_factory=dict)
+
+    @classmethod
+    def get_current(cls, create_if_not_exists=True, **kwargs) -> "TaskContext":
```
nit, I don't think create_if_not_exists and kwargs are needed for this PR. 
Let's remove it.

raulchen reviewed (approved):
LGTM with one last comment
